exec/java-exec/src/main/java/org/apache/drill/exec/planner/DFSDirPartitionLocation.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Class defines a single partition corresponding to a directory in a DFS table.
+ */
+package org.apache.drill.exec.planner;
+
+
+import com.google.common.collect.Lists;
+
+import java.util.Collection;
+import java.util.List;
+
+public class DFSDirPartitionLocation implements PartitionLocation {
+  private final Collection<PartitionLocation> subPartitions;
+  private final String[] dirs;
+
+  public DFSDirPartitionLocation(String[] dirs, Collection<PartitionLocation> subPartitions) {
+    this.subPartitions = subPartitions;
+    this.dirs = dirs;
+  }
+
+  @Override
+  public String getPartitionValue(int index) {
+    assert index < dirs.length;
+    return dirs[index];
+  }
+
+  @Override
+  public String getEntirePartitionLocation() {
+    throw new UnsupportedOperationException("Should not call getEntirePartitionLocation for composite partition location!");
+  }
+
+  @Override
+  public List<PartitionLocation> getPartitionLocationRecursive() {
```

Comment:
I changed this method, such that now it returns list of SimplePartitionLocation.  This method would return all SimplePartitionLocation it consists of. In your example, it would return 4 DFSFilePartitionLocations, if it's called at the DFSDirPartitionLocation corresponding to '2016'. This method is used when we construct a GroupScan after pruning, since only SimplePartitionLocation keeps track the entire path, which is required by a groupscan specification.

The file partition location keeps track of full path (which would be used when created groupscan) and the partition keys. The dir keeps track the nested partition, and the common partition keys.